File: kirun-js/src/engine/function/system/object/ObjectFunctionRepository.ts
import { Namespaces } from '../../../namespaces/Namespaces';
import { Repository } from '../../../Repository';
import { AbstractFunction } from '../../AbstractFunction';
import { Function } from '../../Function';
import { ObjectConvert } from './ObjectConvert';
import { ObjectDeleteKey } from './ObjectDeleteKey';
import { ObjectEntries } from './ObjectEntries';
import { ObjectKeys } from './ObjectKeys';
import { ObjectPutValue } from './ObjectPutValue';
import { ObjectValues } from './ObjectValues';

export class ObjectFunctionRepository implements Repository<Function> {
    private readonly functionObjectsIndex: { [key: string]: AbstractFunction };
    private readonly filterableNames: string[];

    public constructor() {
        this.functionObjectsIndex = {
            ObjectValues: new ObjectValues(),
            ObjectKeys: new ObjectKeys(),
            ObjectEntries: new ObjectEntries(),
            ObjectDeleteKey: new ObjectDeleteKey(),
            ObjectPutValue: new ObjectPutValue(),
            ObjectConvert: new ObjectConvert(),
        };
        this.filterableNames = Object.values(this.functionObjectsIndex).map((e) =>
            e.getSignature().getFullName(),
        );
    }

    public async find(namespace: string, name: string): Promise<Function | undefined> {
        if (namespace != Namespaces.SYSTEM_OBJECT) return Promise.resolve(undefined);

        return Promise.resolve(this.functionObjectsIndex[name]);
    }

    public async filter(name: string): Promise<string[]> {
        return Promise.resolve(
            this.filterableNames.filter((e) => e.toLowerCase().indexOf(name.toLowerCase()) !== -1),
        );
    }
}

File: kirun-py/src/kirun_py/function/system/object/object_function_repository.py
from __future__ import annotations

from typing import Dict, List, Optional

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.function.system.object.object_convert import ObjectConvert
from kirun_py.function.system.object.object_delete_key import ObjectDeleteKey
from kirun_py.function.system.object.object_entries import ObjectEntries
from kirun_py.function.system.object.object_keys import ObjectKeys
from kirun_py.function.system.object.object_put_value import ObjectPutValue
from kirun_py.function.system.object.object_values import ObjectValues
from kirun_py.namespaces.namespaces import Namespaces


class ObjectFunctionRepository:

    def __init__(self) -> None:
        self._function_objects_index: Dict[str, AbstractFunction] = {
            'ObjectValues': ObjectValues(),
            'ObjectKeys': ObjectKeys(),
            'ObjectEntries': ObjectEntries(),
            'ObjectDeleteKey': ObjectDeleteKey(),
            'ObjectPutValue': ObjectPutValue(),
            'ObjectConvert': ObjectConvert(),
        }
        self._filterable_names: List[str] = [
            f.get_signature().get_full_name()
            for f in self._function_objects_index.values()
        ]

    async def find(self, namespace: str, name: str) -> Optional[AbstractFunction]:
        if namespace != Namespaces.SYSTEM_OBJECT:
            return None
        return self._function_objects_index.get(name)

    async def filter(self, name: str) -> List[str]:
        lower_name = name.lower()
        return [e for e in self._filterable_names if lower_name in e.lower()]
